### PR TITLE
Optimize Token Classification models for TPU

### PR DESCRIPTION
**Pull request by ibraheem-moosa**
As per the XLA [documentation](https://github.com/pytorch/xla/blob/master/TROUBLESHOOTING.md#known-performance-caveats) XLA cannot handle masked indexing well. So token classification
models for BERT and others use an implementation based on `torch.where`. This implementation
works well on TPU. 

ALBERT, ELECTRA and LayoutLM token classification model uses the masked indexing which 
causes performance issues on TPU. 

This PR fixes this issue by following the BERT implementation.

Relevant code in [BERT](https://github.com/huggingface/transformers/blob/c4e1586db8ef6b4102016fc5cb038940fde45325/src/transformers/models/bert/modeling_bert.py#L1741) 

# What does this PR do?

Fixes # (issue)

## Before submitting
- [ ] This PR fixes a typo or improves the docs (you can dismiss the other checks if that's the case).
- [ ] Did you read the [contributor guideline](https://github.com/huggingface/transformers/blob/master/CONTRIBUTING.md#start-contributing-pull-requests),
      Pull Request section?
- [ ] Was this discussed/approved via a Github issue or the [forum](https://discuss.huggingface.co/)? Please add a link
      to it if that's the case.
- [ ] Did you make sure to update the documentation with your changes? Here are the
      [documentation guidelines](https://github.com/huggingface/transformers/tree/master/docs), and
      [here are tips on formatting docstrings](https://github.com/huggingface/transformers/tree/master/docs#writing-source-documentation).
- [ ] Did you write any new necessary tests?

## Who can review?

Anyone in the community is free to review the PR once the tests have passed. Feel free to tag
members/contributors who may be interested in your PR.

### CODE REVIEW

**Comment by LysandreJik** [comment]
Hello, thank you for your PR! Would you happen to have performance results we can take a look at to see the improvement your PR offers?

**Comment by ibraheem-moosa** [comment]
Sure. I came into this issue while running experiments for our paper. I cannot make that code public yet. I will make a colab notebook illustrating the issue.

**Comment by ibraheem-moosa** [comment]
Hi. Sorry for being so late. I have prepared a colab notebook showing the speedup [here](https://colab.research.google.com/drive/1Y9f3BWkTeQS7lFJSGXjcCULURYAJ1NZw?usp=sharing).  With this PR we can improve ALBERT Token Classification model training time from 27.5 minutes to 3minutes.

**Comment by LysandreJik** [comment]
Thank you for sharing, I have requested access to the doc

**Comment by ibraheem-moosa** [comment]
I have updated the permission.

**Comment by LysandreJik** [comment]
Thank you, this looks good! @sgugger, @mfuntowicz, I think you're the most experienced with torch operations - do you have some feedback for this PR?

**Review by sgugger — APPROVED** [review]
I think this is a welcome change, but just to be sure, can you run some benchmark on GPU as well? I don't think it changes anything there but I'd like to be 100% sure.

**Comment by ibraheem-moosa** [comment]
Sure. I will run the code on GPU too :)

**Comment by ibraheem-moosa** [comment]
Here is the [colab notebook](https://colab.research.google.com/drive/1OTBfHNt-ZGqFlz1kFvFEq-oCBRYFEuLw?usp=sharing) comparing the execution of the original implementation and the patched implementation on GPU. As expected, there is no performance degradation on GPU.

**Comment by sgugger** [comment]
Thanks for checking! The failing test has been fixed on master so this is good to merge for me.

**Comment by LysandreJik** [comment]
Thanks for your work @ibraheem-moosa, super nice addition!
